Enumerize tracepoint categories as described in #314

syogaraj opened the pull request:
1. Used `trace-cmd list -e` for the list of tracepoint categories.
2. Tracepoint now has its own file (`tracepoint.rs`). Hence, removed `TracepointOpts` from `program.rs` and placed it in `tracepoint.rs`.

danielocfb commented on libbpf-rs/src/tracepoint.rs:
```diff
@@ -0,0 +1,492 @@
+    Xen,
+    /// xHCI Host Controller events.
+    XhciHcd,
+    /// Unknown,
+    Unknown,
```
The way I see it this variant doesn't make much sense here. This is user supplied data and the kernel doesn't understand `unknown`.
I think we should replace this with a variant that allows for passing in an arbitrary string, similar to what was possible before. This way, even if a new variant is added in the kernel but hasn't been hoked up to `libbpf-rs` yet, users would still be able to attach the tracepoint.

syogaraj replied:
Understood. Updated with this change.

chenhengqi reviewed (commented):
Looks good overall. But first commit is incomplete, you can't not refer to TracepointCategory::Syscalls defined in the second commit.

danielocfb commented on libbpf-rs/src/tracepoint.rs:
```diff
@@ -0,0 +1,493 @@
+#[non_exhaustive]
+#[repr(u32)]
+#[derive(Debug)]
+pub enum TracepointCategory {
```
You are saying remove everything that is arch or driver specific? I have no strong preference, but to me that seems like a moving target: over time something may become generic. It also seems somewhat arbitrary and ultimately we just provide possible values, but runtime detection by the kernel should have the last say anyway. So personally I guess I'd err on the side of making more values available.

chenhengqi replied:
This list seems taken from some random running machine, especially the amdcpu thing.

There are tens of arch and thousands of drivers, trying to keep them here also a moving target.

danielocfb replied:
I see. In that case, fair enough, perhaps it's best to trim it down. @syogaraj what do you think? Let's only provide variants for core categories?

syogaraj replied:
> This list seems taken from some random running machine, especially the amdcpu thing.

This list is taken from a machine that is running on Intel CPU (11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz) with kernel v6.11.0-17. Don't know why amdcpu is listed. 

> what do you think? Let's only provide variants for core categories?

Even with numerous non-core variants, users can still utilize the `Custom` category as needed for those categories that are missing. Moreover, reducing the list now would entail extra work and additional checks, which we can avoid. I'm comfortable proceeding with the current list. Your thoughts on this?

chenhengqi replied:
Nope, this should be easy. Just grep `TRACE_SYSTEM` from https://github.com/torvalds/linux/tree/master/include/trace/events

chenhengqi replied:
I am with the list above. An allyesconfig should compile all this in.

syogaraj replied:
I've pushed the changes with the tracepoint categories shared by @d-e-s-o. However, the "syscalls" and "bpf_trace_event" categories were not in the generated list, so I added these two as extras. I hope this is fine.

danielocfb commented:
I see there are some conflicts with the merged https://github.com/libbpf/libbpf-rs/pull/1115. I believe they should be fairly mechanical to resolve, though. Sorry about that, but we'd have conflicts either way.

syogaraj commented:
> > But first commit is incomplete, you can't not refer to TracepointCategory::Syscalls defined in the second commit.
> 
> Yeah, that's a good catch. Can you please fix that up and make sure everything builds at each point, @syogaraj ?

Missed this @chenhengqi. fixed them now. @danielocfb


> I see there are some conflicts with the merged #1115. I believe they should be fairly mechanical to resolve, though. Sorry about that, but we'd have conflicts either way.

@danielocfb, Resolved the merge conflict from #1115 and moved `RawTracepointOpts` to `tracepoint` module.

danielocfb commented:
Thanks! Will take another look tomorrow.

danielocfb commented on libbpf-rs/src/tracepoint.rs:
```diff
@@ -52,3 +52,466 @@ impl From<RawTracepointOpts> for libbpf_sys::bpf_raw_tracepoint_opts {
 }
+
+
+/// Represents categories of Linux kernel tracepoints.
```
Let's mark this type as non-exhaustive similar to what we do for the `MapType`. https://github.com/libbpf/libbpf-rs/blob/9d617b054515be03b7dc37897e511a4574197de1/libbpf-rs/src/map.rs#L1211-L1216

syogaraj replied:
@chenhengqi suggested that we remove this non_exhaustive check [earlier](https://github.com/libbpf/libbpf-rs/pull/1114#discussion_r1984519742). Added this again.

syogaraj commented on libbpf-rs/src/tracepoint.rs:
```diff
@@ -52,3 +52,466 @@ impl From<RawTracepointOpts> for libbpf_sys::bpf_raw_tracepoint_opts {
+///
+/// This enum provides a list of tracepoint categories that can be used with
+/// BPF programs to attach to various kernel events.
+#[derive(Debug)]
```
added..

d-e-s-o reviewed (approved):
Looks good to me now.

d-e-s-o commented on libbpf-rs/src/tracepoint.rs:
```diff
@@ -52,3 +52,506 @@ impl From<RawTracepointOpts> for libbpf_sys::bpf_raw_tracepoint_opts {
+#[derive(Debug, Clone)]
+pub enum TracepointCategory {
+    /// 9P protocol events.
+    P9,
```
Haha, okay, I see why you chose that now.
